File: server/src/autolaunch.ts
import AutoLaunch from "auto-launch";
import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { homedir } from "node:os";
import { fileURLToPath } from "node:url";

const APP_NAME = "Remote Media Control";

function startupDir(): string {
  const base =
    process.platform === "win32"
      ? process.env.APPDATA ?? join(homedir(), "AppData", "Roaming")
      : process.env.XDG_CONFIG_HOME ?? join(homedir(), ".config");
  return join(base, "remote-media-control");
}

function writeWindowsWrapper(): string {
  const here = dirname(fileURLToPath(import.meta.url));
  const entry = resolve(here, "index.js");
  const wrapper = join(startupDir(), "start.vbs");
  const node = process.execPath.replace(/"/g, '""');
  const script = `Set sh = CreateObject("WScript.Shell")\nsh.Run """${node}"" ""${entry.replace(/"/g, '""')}""", 0, False\n`;
  mkdirSync(dirname(wrapper), { recursive: true });
  writeFileSync(wrapper, script);
  return wrapper;
}

function launcherPath(): string {
  if (process.platform === "win32") return writeWindowsWrapper();
  return process.execPath;
}

function createLauncher(): AutoLaunch {
  return new AutoLaunch({
    name: APP_NAME,
    path: launcherPath(),
    isHidden: true
  });
}

export async function setAutoLaunch(enabled: boolean): Promise<void> {
  const launcher = createLauncher();
  const already = await launcher.isEnabled();
  if (enabled && !already) await launcher.enable();
  if (!enabled && already) await launcher.disable();
}

export async function isAutoLaunchEnabled(): Promise<boolean> {
  try {
    return await createLauncher().isEnabled();
  } catch {
    return false;
  }
}

File: server/src/autolaunch.rs
pub fn set_auto_launch(enabled: bool) -> Result<(), String> {
    let exe = std::env::current_exe().map_err(|e| e.to_string())?;
    platform::set_auto_launch(enabled, &exe)
}

#[cfg(target_os = "windows")]
mod platform {
    use std::path::PathBuf;
    use winreg::enums::*;
    use winreg::RegKey;

    const APP_NAME: &str = "Sofamote";
    const RUN_KEY: &str = "Software\\Microsoft\\Windows\\CurrentVersion\\Run";

    pub fn set_auto_launch(enabled: bool, exe: &PathBuf) -> Result<(), String> {
        let hkcu = RegKey::predef(HKEY_CURRENT_USER);
        let run = hkcu
            .open_subkey_with_flags(RUN_KEY, KEY_SET_VALUE | KEY_QUERY_VALUE)
            .map_err(|e| e.to_string())?;

        if enabled {
            let vbs_path = write_vbs_wrapper(exe).map_err(|e| e.to_string())?;
            let cmd = format!("wscript.exe \"{}\"", vbs_path.display());
            run.set_value(APP_NAME, &cmd).map_err(|e| e.to_string())
        } else {
            run.delete_value(APP_NAME).or_else(|_| Ok(()))
        }
    }

    fn write_vbs_wrapper(exe: &PathBuf) -> std::io::Result<PathBuf> {
        let config_dir = dirs::config_dir().unwrap().join("sofamote");
        std::fs::create_dir_all(&config_dir)?;
        let vbs = config_dir.join("start.vbs");
        let script = format!(
            "Set sh = CreateObject(\"WScript.Shell\")\nsh.Run \"\"\"{}\"\"\", 0, False\n",
            exe.display()
        );
        std::fs::write(&vbs, script)?;
        Ok(vbs)
    }
}

#[cfg(target_os = "linux")]
mod platform {
    use std::path::PathBuf;

    const DESKTOP_FILE: &str = "sofamote.desktop";

    pub fn set_auto_launch(enabled: bool, exe: &PathBuf) -> Result<(), String> {
        let autostart = dirs::config_dir().ok_or("no config dir")?.join("autostart");
        let desktop = autostart.join(DESKTOP_FILE);
        if enabled {
            std::fs::create_dir_all(&autostart).map_err(|e| e.to_string())?;
            let content = format!(
                "[Desktop Entry]\nType=Application\nName=Sofamote\nExec={}\nHidden=false\nNoDisplay=false\nX-GNOME-Autostart-enabled=true\n",
                exe.display()
            );
            std::fs::write(&desktop, content).map_err(|e| e.to_string())
        } else {
            std::fs::remove_file(&desktop).or_else(|_| Ok(()))
        }
    }
}

#[cfg(target_os = "macos")]
mod platform {
    use std::path::PathBuf;

    const PLIST_FILE: &str = "com.sofamote.plist";

    pub fn set_auto_launch(enabled: bool, exe: &PathBuf) -> Result<(), String> {
        let agents_dir = dirs::home_dir()
            .ok_or("no home dir")?
            .join("Library")
            .join("LaunchAgents");
        let plist = agents_dir.join(PLIST_FILE);
        if enabled {
            std::fs::create_dir_all(&agents_dir).map_err(|e| e.to_string())?;
            let content = format!(
                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n\
                 <!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n\
                 <plist version=\"1.0\">\n\
                 <dict>\n\
                   <key>Label</key><string>com.sofamote</string>\n\
                   <key>ProgramArguments</key>\n\
                   <array><string>{}</string></array>\n\
                   <key>RunAtLoad</key><true/>\n\
                 </dict>\n\
                 </plist>\n",
                exe.display()
            );
            std::fs::write(&plist, content).map_err(|e| e.to_string())
        } else {
            std::fs::remove_file(&plist).or_else(|_| Ok(()))
        }
    }
}

#[cfg(not(any(target_os = "windows", target_os = "linux", target_os = "macos")))]
mod platform {
    use std::path::PathBuf;
    pub fn set_auto_launch(_enabled: bool, _exe: &PathBuf) -> Result<(), String> {
        Err("auto-launch not supported on this platform".into())
    }
}
